Lots of missing npm packages errors when pnpm 3.8.1 is globally installed

Running `mvn install` on a 15 alpha 10 project results in errors like
```
Error: Cannot find module 'mkdirp'
Require stack:
- /.../project-folder/webpack.generated.js
- /.../project-folder/webpack.config.js
```

```
pnpm --version
3.8.1
```

If there is a minimum version of `pnpm` required, there should be a version check like with `npm`

Upgrading globally installed pnpm to 4.0.0 resolves the issue

--- a/flow-migration/src/main/java/com/vaadin/flow/migration/Migration.java
+++ b/flow-migration/src/main/java/com/vaadin/flow/migration/Migration.java
@@ -309,14 +309,13 @@ public class Migration {
     }
 
     private List<String> getFrontendInstallToolExec() {
-        FrontendUtils.ensurePnpm(configuration.getBaseDirectory().getPath(),
-                !configuration.isPnpmDisabled());
-
         List<String> executable;
         if (configuration.isPnpmDisabled()) {
             executable = FrontendUtils.getNpmExecutable(
                     configuration.getBaseDirectory().getPath());
         } else {
+            FrontendUtils
+                    .ensurePnpm(configuration.getBaseDirectory().getPath());
             executable = FrontendUtils.getPnpmExecutable(
                     configuration.getBaseDirectory().getPath());
         }

--- a/flow-server/src/main/java/com/vaadin/flow/server/Constants.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/Constants.java
@@ -267,6 +267,9 @@ public final class Constants implements Serializable {
     public static final int SHOULD_WORK_NPM_MAJOR_VERSION = 5;
     public static final int SHOULD_WORK_NPM_MINOR_VERSION = 5;
 
+    public static final int SUPPORTED_PNPM_MAJOR_VERSION = 4;
+    public static final int SUPPORTED_PNPM_MINOR_VERSION = 4;
+
     /**
      * Property boolean for marking stats.json to be fetched from external
      * location.

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -777,30 +777,7 @@ public class FrontendUtils {
      *            project root folder.
      */
     public static void ensurePnpm(String baseDir) {
-        final List<String> pnpm = getPnpmExecutable(baseDir, false);
-        boolean installPnpmLocally = true;
-        if (!pnpm.isEmpty()) {
-            // Check if (globally or locally installed) pnpm is new enough. If
-            // not, install a supported version locally.
-            try {
-                List<String> versionCmd = new ArrayList<>(pnpm);
-                versionCmd.add("--version");
-                FrontendVersion pnpmVersion = getVersion("pnpm", versionCmd);
-                if (isVersionAtLeast(pnpmVersion, SUPPORTED_PNPM_VERSION)) {
-                    installPnpmLocally = false;
-                } else {
-                    getLogger().warn(String.format(
-                            "installed pnpm ('%s', version %s) is too old, installing supported version locally",
-                            String.join(" ", pnpm),
-                            pnpmVersion.getFullVersion()));
-                }
-            } catch (UnknownVersionException e) {
-                getLogger().warn(
-                        "Error checking pnpm version, installing pnpm locally",
-                        e);
-            }
-        }
-        if (installPnpmLocally) {
+        if (isPnpmTooOldOrAbsent(baseDir)) {
             // copy the current content of package.json file to a temporary
             // location
             File packageJson = new File(baseDir, "package.json");
@@ -835,6 +812,31 @@ public class FrontendUtils {
                 tempFile.delete();
             }
         }
+    }
+
+    private static boolean isPnpmTooOldOrAbsent(String baseDir) {
+        final List<String> pnpmCommand = getPnpmExecutable(baseDir, false);
+        if (!pnpmCommand.isEmpty()) {
+            // check whether globally or locally installed pnpm is new enough
+            try {
+                List<String> versionCmd = new ArrayList<>(pnpmCommand);
+                versionCmd.add("--version");
+                FrontendVersion pnpmVersion = getVersion("pnpm", versionCmd);
+                if (isVersionAtLeast(pnpmVersion, SUPPORTED_PNPM_VERSION)) {
+                    return false;
+                } else {
+                    getLogger().warn(String.format(
+                            "installed pnpm ('%s', version %s) is too old, installing supported version locally",
+                            String.join(" ", pnpmCommand),
+                            pnpmVersion.getFullVersion()));
+                }
+            } catch (UnknownVersionException e) {
+                getLogger().warn(
+                        "Error checking pnpm version, installing pnpm locally",
+                        e);
+            }
+        }
+        return true;
     }
 
     static void checkForFaultyNpmVersion(FrontendVersion npmVersion) {

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -262,7 +262,11 @@ public class FrontendUtils {
             Constants.SHOULD_WORK_NPM_MAJOR_VERSION,
             Constants.SHOULD_WORK_NPM_MINOR_VERSION);
 
-    private static FrontendToolsLocator frontendToolsLocator = new FrontendToolsLocator();
+    private static final FrontendVersion SUPPORTED_PNPM_VERSION = new FrontendVersion(
+            Constants.SUPPORTED_PNPM_MAJOR_VERSION,
+            Constants.SUPPORTED_PNPM_MINOR_VERSION);
+
+    static FrontendToolsLocator frontendToolsLocator = new FrontendToolsLocator();
 
     private static String operatingSystem = null;
 
@@ -350,8 +354,12 @@ public class FrontendUtils {
      * @see #getPnpmExecutable(String, boolean)
      */
     public static List<String> getPnpmExecutable(String baseDir) {
-        ensurePnpm(baseDir, true);
-        return getPnpmExecutable(baseDir, true);
+        ensurePnpm(baseDir);
+        List<String> pnpmCommand = getPnpmExecutable(baseDir, true);
+        if (!pnpmCommand.isEmpty()) {
+            pnpmCommand.add("--shamefully-hoist=true");
+        }
+        return pnpmCommand;
     }
 
     /**
@@ -390,9 +398,6 @@ public class FrontendUtils {
                         .map(File::getPath).map(Collections::singletonList)
                         .orElse(Collections.emptyList()));
             }
-        }
-        if (!returnCommand.isEmpty()) {
-            returnCommand.add("--shamefully-hoist=true");
         }
         return returnCommand;
     }
@@ -770,11 +775,32 @@ public class FrontendUtils {
      *
      * @param baseDir
      *            project root folder.
-     * @param ensure
-     *            whether pnpm tool should be installed if it's absent
      */
-    public static void ensurePnpm(String baseDir, boolean ensure) {
-        if (ensure && getPnpmExecutable(baseDir, false).isEmpty()) {
+    public static void ensurePnpm(String baseDir) {
+        final List<String> pnpm = getPnpmExecutable(baseDir, false);
+        boolean installPnpmLocally = true;
+        if (!pnpm.isEmpty()) {
+            // Check if (globally or locally installed) pnpm is new enough. If
+            // not, install a supported version locally.
+            try {
+                List<String> versionCmd = new ArrayList<>(pnpm);
+                versionCmd.add("--version");
+                FrontendVersion pnpmVersion = getVersion("pnpm", versionCmd);
+                if (isVersionAtLeast(pnpmVersion, SUPPORTED_PNPM_VERSION)) {
+                    installPnpmLocally = false;
+                } else {
+                    getLogger().warn(String.format(
+                            "installed pnpm ('%s', version %s) is too old, installing supported version locally",
+                            String.join(" ", pnpm),
+                            pnpmVersion.getFullVersion()));
+                }
+            } catch (UnknownVersionException e) {
+                getLogger().warn(
+                        "Error checking pnpm version, installing pnpm locally",
+                        e);
+            }
+        }
+        if (installPnpmLocally) {
             // copy the current content of package.json file to a temporary
             // location
             File packageJson = new File(baseDir, "package.json");

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -266,7 +266,7 @@ public class FrontendUtils {
             Constants.SUPPORTED_PNPM_MAJOR_VERSION,
             Constants.SUPPORTED_PNPM_MINOR_VERSION);
 
-    static FrontendToolsLocator frontendToolsLocator = new FrontendToolsLocator();
+    private static FrontendToolsLocator frontendToolsLocator = new FrontendToolsLocator();
 
     private static String operatingSystem = null;
 

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/FrontendUtils.java
@@ -747,7 +747,7 @@ public class FrontendUtils {
         try {
             List<String> nodeVersionCommand = new ArrayList<>();
             nodeVersionCommand.add(FrontendUtils.getNodeExecutable(baseDir));
-            nodeVersionCommand.add("--version");
+            nodeVersionCommand.add("--version"); //NOSONAR
             FrontendVersion nodeVersion = getVersion("node",
                     nodeVersionCommand);
             validateToolVersion("node", nodeVersion, SUPPORTED_NODE_VERSION,
@@ -759,7 +759,7 @@ public class FrontendUtils {
         try {
             List<String> npmVersionCommand = new ArrayList<>(
                     FrontendUtils.getNpmExecutable(baseDir));
-            npmVersionCommand.add("--version");
+            npmVersionCommand.add("--version"); //NOSONAR
             FrontendVersion npmVersion = getVersion("npm", npmVersionCommand);
             validateToolVersion("npm", npmVersion, SUPPORTED_NPM_VERSION,
                     SHOULD_WORK_NPM_VERSION);
@@ -820,7 +820,7 @@ public class FrontendUtils {
             // check whether globally or locally installed pnpm is new enough
             try {
                 List<String> versionCmd = new ArrayList<>(pnpmCommand);
-                versionCmd.add("--version");
+                versionCmd.add("--version"); //NOSONAR
                 FrontendVersion pnpmVersion = getVersion("pnpm", versionCmd);
                 if (isVersionAtLeast(pnpmVersion, SUPPORTED_PNPM_VERSION)) {
                     return false;

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -22,13 +22,19 @@ import java.io.IOException;
 import java.io.PrintStream;
 import java.io.UnsupportedEncodingException;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.util.Arrays;
 import java.util.List;
+import java.util.Optional;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import org.apache.commons.io.FileUtils;
 import org.apache.commons.io.IOUtils;
 import org.junit.Assert;
+import org.junit.Assume;
 import org.junit.Before;
 import org.junit.Rule;
 import org.junit.Test;
@@ -61,6 +67,11 @@ public class FrontendUtilsTest {
     public static final String NPM_CLI_STRING = Stream
             .of("node", "node_modules", "npm", "bin", "npm-cli.js")
             .collect(Collectors.joining(File.separator));
+
+    public static final String PNPM_INSTALL_LOCATION = Stream
+            .of("node_modules","pnpm","bin","pnpm.js")
+            .collect(Collectors.joining(File.separator));
+
     @Rule
     public ExpectedException exception = ExpectedException.none();
 
@@ -270,17 +281,6 @@ public class FrontendUtilsTest {
         Assert.assertNull(statsAssetsByChunkName);
     }
 
-    @Test
-    public void ensurePnpm_doNotEnsure() {
-        List<String> executable = FrontendUtils.getPnpmExecutable(baseDir,
-                false);
-
-        FrontendUtils.ensurePnpm(baseDir, false);
-
-        Assert.assertEquals(executable,
-                FrontendUtils.getPnpmExecutable(baseDir, false));
-    }
-
     /**
      * This test doesn't do anything if pnpm is already installed (globally)
      * which is true e.g. for or CI servers (TC/bender).
@@ -288,28 +288,26 @@ public class FrontendUtilsTest {
     @Test
     public void ensurePnpm_requestInstall_keepPackageJson_removePackageLock_ignoredPnpmExists_localPnpmIsRemoved()
             throws IOException {
-        List<String> executable = FrontendUtils.getPnpmExecutable(baseDir,
-                false);
-        if (executable.isEmpty()) {
-            File packageJson = new File(baseDir, "package.json");
-            FileUtils.writeStringToFile(packageJson, "{}",
-                    StandardCharsets.UTF_8);
+        Assume.assumeTrue(FrontendUtils.getPnpmExecutable(baseDir,
+                false).isEmpty());
+        File packageJson = new File(baseDir, "package.json");
+        FileUtils.writeStringToFile(packageJson, "{}",
+                StandardCharsets.UTF_8);
 
-            File packageLockJson = new File(baseDir, "package-lock.json");
-            FileUtils.writeStringToFile(packageLockJson, "{}",
-                    StandardCharsets.UTF_8);
+        File packageLockJson = new File(baseDir, "package-lock.json");
+        FileUtils.writeStringToFile(packageLockJson, "{}",
+                StandardCharsets.UTF_8);
 
-            FrontendUtils.ensurePnpm(baseDir, true);
-            Assert.assertFalse(
-                    FrontendUtils.getPnpmExecutable(baseDir, false).isEmpty());
+        FrontendUtils.ensurePnpm(baseDir);
+        Assert.assertFalse(
+                FrontendUtils.getPnpmExecutable(baseDir, false).isEmpty());
 
-            // locally installed pnpm (via npm/pnpm) is removed
-            Assert.assertFalse(new File("node_modules/pnpm").exists());
+        // locally installed pnpm (via npm/pnpm) is removed
+        Assert.assertFalse(new File("node_modules/pnpm").exists());
 
-            Assert.assertEquals("{}", FileUtils.readFileToString(packageJson,
-                    StandardCharsets.UTF_8));
-            Assert.assertFalse(packageLockJson.exists());
-        }
+        Assert.assertEquals("{}", FileUtils.readFileToString(packageJson,
+                StandardCharsets.UTF_8));
+        Assert.assertFalse(packageLockJson.exists());
     }
 
     @Test
@@ -319,6 +317,48 @@ public class FrontendUtilsTest {
         Assert.assertTrue(executable.contains("--shamefully-hoist=true"));
         Assert.assertTrue(
                 executable.stream().anyMatch(cmd -> cmd.contains("pnpm")));
+    }
+
+    // #7219
+    @Test
+    public void ensurePnpm_globalPnpmIsTooOld_localPnpmIsInstalled()
+            throws IOException {
+        // this unit test must be run on a Unix-like OS
+        Assume.assumeFalse(FrontendUtils.isWindows());
+
+        FrontendToolsLocator defaultFrontendToolsLocator = FrontendUtils.frontendToolsLocator;
+        try {
+            // given: an existing globally installed version of pnpm that is too
+            // old
+            FrontendUtils.frontendToolsLocator = new FrontendToolsLocator() {
+                private final Path oldPnpm = Files.createTempFile("pnpm", "old",
+                        PosixFilePermissions.asFileAttribute(
+                                PosixFilePermissions.fromString("rwxrwxrwx")));
+                {
+                    Files.write(oldPnpm, Arrays.asList("#!/bin/sh",
+                            "if [ $1 = '--version' ]; then echo '3.8.1'; fi"),
+                            StandardCharsets.UTF_8);
+                }
+
+                public Optional<File> tryLocateTool(String toolName) {
+                    return "pnpm".equals(toolName)
+                            ? Optional.of(oldPnpm.toFile())
+                            : super.tryLocateTool(toolName);
+                }
+            };
+
+            // when
+            FrontendUtils.ensurePnpm(baseDir);
+
+            // then: pnpm is installed locally
+            List<String> pnpmExecutable = FrontendUtils
+                    .getPnpmExecutable(baseDir, false);
+            Assert.assertTrue(pnpmExecutable.size() > 1);
+            Assert.assertTrue(
+                    pnpmExecutable.get(1).contains(PNPM_INSTALL_LOCATION));
+        } finally {
+            FrontendUtils.frontendToolsLocator = defaultFrontendToolsLocator;
+        }
     }
 
     private VaadinService setupStatsAssetMocks(String statsFile)

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/FrontendUtilsTest.java
@@ -319,48 +319,6 @@ public class FrontendUtilsTest {
                 executable.stream().anyMatch(cmd -> cmd.contains("pnpm")));
     }
 
-    // #7219
-    @Test
-    public void ensurePnpm_globalPnpmIsTooOld_localPnpmIsInstalled()
-            throws IOException {
-        // this unit test must be run on a Unix-like OS
-        Assume.assumeFalse(FrontendUtils.isWindows());
-
-        FrontendToolsLocator defaultFrontendToolsLocator = FrontendUtils.frontendToolsLocator;
-        try {
-            // given: an existing globally installed version of pnpm that is too
-            // old
-            FrontendUtils.frontendToolsLocator = new FrontendToolsLocator() {
-                private final Path oldPnpm = Files.createTempFile("pnpm", "old",
-                        PosixFilePermissions.asFileAttribute(
-                                PosixFilePermissions.fromString("rwxrwxrwx")));
-                {
-                    Files.write(oldPnpm, Arrays.asList("#!/bin/sh",
-                            "if [ $1 = '--version' ]; then echo '3.8.1'; fi"),
-                            StandardCharsets.UTF_8);
-                }
-
-                public Optional<File> tryLocateTool(String toolName) {
-                    return "pnpm".equals(toolName)
-                            ? Optional.of(oldPnpm.toFile())
-                            : super.tryLocateTool(toolName);
-                }
-            };
-
-            // when
-            FrontendUtils.ensurePnpm(baseDir);
-
-            // then: pnpm is installed locally
-            List<String> pnpmExecutable = FrontendUtils
-                    .getPnpmExecutable(baseDir, false);
-            Assert.assertTrue(pnpmExecutable.size() > 1);
-            Assert.assertTrue(
-                    pnpmExecutable.get(1).contains(PNPM_INSTALL_LOCATION));
-        } finally {
-            FrontendUtils.frontendToolsLocator = defaultFrontendToolsLocator;
-        }
-    }
-
     private VaadinService setupStatsAssetMocks(String statsFile)
             throws IOException {
         String stats = IOUtils.toString(FrontendUtilsTest.class.getClassLoader()

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunPnpmInstallTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunPnpmInstallTest.java
@@ -25,8 +25,7 @@ public class TaskRunPnpmInstallTest extends TaskRunNpmInstallTest {
     @Before
     public void setUp() throws IOException {
         super.setUp();
-        FrontendUtils.ensurePnpm(getNodeUpdater().npmFolder.getAbsolutePath(),
-                true);
+        FrontendUtils.ensurePnpm(getNodeUpdater().npmFolder.getAbsolutePath());
     }
 
     @Override

